// @squidcloud/common/dist/integration.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDataIntegrationType = exports.isDataIntegration = exports.IntegrationType = void 0;
var IntegrationType;
(function (IntegrationType) {
    IntegrationType["built_in_db"] = "built_in_db";
    IntegrationType["mongo"] = "mongo";
    IntegrationType["mysql"] = "mysql";
    IntegrationType["mssql"] = "mssql";
    IntegrationType["postgres"] = "postgres";
    IntegrationType["cockroach"] = "cockroach";
    IntegrationType["api"] = "api";
    IntegrationType["graphql"] = "graphql";
})(IntegrationType = exports.IntegrationType || (exports.IntegrationType = {}));
function isDataIntegration(integration) {
    return isDataIntegrationType(integration.type);
}
exports.isDataIntegration = isDataIntegration;
function isDataIntegrationType(type) {
    return [
        IntegrationType.built_in_db,
        IntegrationType.mongo,
        IntegrationType.mysql,
        IntegrationType.mssql,
        IntegrationType.postgres,
        IntegrationType.cockroach,
    ].includes(type);
}
exports.isDataIntegrationType = isDataIntegrationType;
