// @squidcloud/common/dist/secret.schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateNewApiKeyRequestSchema = exports.DeleteApiKeyRequestSchema = exports.DeleteCustomSecretRequestSchema = exports.SetCustomSecretRequestSchema = void 0;
exports.SetCustomSecretRequestSchema = {
    type: 'object',
    required: ['key', 'value'],
    properties: {
        key: { type: 'string', pattern: '[a-zA-Z0-9-_+]', nullable: false },
        value: { type: 'string', nullable: false },
    },
};
exports.DeleteCustomSecretRequestSchema = {
    type: 'object',
    required: ['key'],
    properties: {
        key: { type: 'string', pattern: '[a-zA-Z0-9-_]+', nullable: false, isSecret: {} },
    },
};
exports.DeleteApiKeyRequestSchema = {
    type: 'object',
    required: ['key'],
    properties: {
        key: { type: 'string', pattern: '[a-zA-Z0-9-_]+', nullable: false, isSecret: { collection: 'apiKeys' } },
    },
};
exports.GenerateNewApiKeyRequestSchema = {
    type: 'object',
    required: ['key'],
    properties: {
        key: { type: 'string', pattern: '[a-zA-Z0-9-_]+', nullable: false, isSecret: {} },
    },
};
